src/libraries/System.Private.CoreLib/src/System/IO/Path.cs

```diff
@@ -926,5 +926,56 @@ private static string GetRelativePath(string relativeTo, string path, StringComp
         /// </summary>
         public static bool EndsInDirectorySeparator(string path) => PathInternal.EndsInDirectorySeparator(path);
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A string containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        public static string RemoveRedundantSegments(string path)
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+
+            if (PathInternal.IsEffectivelyEmpty(path))
+                throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
+
+            return RemoveRedundantSegments(path.AsSpan());
+        }
+
+        /// <summary>
+        /// Removes any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <returns><paramref name="path" /> without any redundant segments.</returns>
+        public static string RemoveRedundantSegments(ReadOnlySpan<char> path)
+        {
+            if (PathInternal.IsEffectivelyEmpty(path))
+                throw new ArgumentException(SR.Arg_PathEmpty, nameof(path));
+
+            Span<char> destination = stackalloc char[path.Length];
+
+            PathInternal.TryRemoveRedundantSegments(path, destination, out int charsWritten);
+
+            return destination.Slice(0, charsWritten).ToString();
+        }
+
+        /// <summary>
+        /// Tries to remove any redundant segments found in the specified path.
+        /// </summary>
+        /// <param name="path">A read-only span containing an absolute or relative path that may or may not contain redundant segments.</param>
+        /// <param name="destination">When the method returns <see langword="true" />, a span containing <paramref name="path" /> without any redundant segments; when the method returns <see langword="false" />, this parameter remains unmodified.</param>
+        /// <param name="charsWritten">When the method returns <see langword="true" />, contains the total number of characters written in <paramref name="destination" />; when the method returns <see langword="false" />, the value is zero.</param>
+        /// <returns><see langword="true" /> if the operation succeeds; <see langword="false" /> if the operation fails.</returns>
+        /// <remarks>The method returns `false` when the path is empty, when the destination span's length is less than the path's length, or when attempting to remove the relative segments results in a failure.</remarks>
+        public static bool TryRemoveRedundantSegments(ReadOnlySpan<char> path, Span<char> destination, out int charsWritten)
```

Comment:
It looks like this API always "shrinks" the input (or leaves it as is). Does it make sense to provide an overload that does an in-place update that takes a `Span<char>` or will most people only have `ROSpan<char>`?

Reply:
I followed Jeremy's suggestion in his other comment to pass a ValueStringBuilder from all the callers, and ensure it's initialized with the same length as the user's path. That way the underlying span will never have to be increased in size.